src/Core/Content/Property/PropertyGroupCollection.php

```diff
@@ -47,20 +47,30 @@ public function sortByPositions(): void
 
     public function sortByConfig(): void
     {
-        /** @var Entity $group */
         foreach ($this->elements as $group) {
-            $options = $group->get('options');
-            if (!$options instanceof EntityCollection) {
+            $options = $group->getOptions();
+            if (!$options instanceof PropertyGroupOptionCollection) {
                 continue;
             }
 
-            $options->sort(static function (Entity $a, Entity $b) use ($group) {
-                if ($group->get('sortingType') === PropertyGroupDefinition::SORTING_TYPE_ALPHANUMERIC) {
-                    return strnatcmp((string) $a->getTranslation('name'), (string) $b->getTranslation('name'));
+            $columns = [];
+            $entities = [];
+
+            $sortingType = $group->getSortingType();
+
+            foreach ($options->getIterator() as $option) {
+                if ($sortingType === PropertyGroupDefinition::SORTING_TYPE_ALPHANUMERIC) {
+                    $columns[] = (string) ($option->getTranslation('name') ?? '');
+                } else {
+                    $columns[] = (int) ($option->getTranslation('position') ?? $option->getPosition() ?? 0);
                 }
 
-                return ($a->getTranslation('position') ?? $a->get('position') ?? 0) <=> ($b->getTranslation('position') ?? $b->get('position') ?? 0);
-            });
+                $entities[] = $option;
+            }
+
+            array_multisort($columns, \SORT_ASC, \SORT_NATURAL, $entities);
+
+            $group->setOptions(new PropertyGroupOptionCollection($entities));
```

Comment:
When passing $entities to a Collection, we always need to validateType, which is unnecessary on many occasions, so we can add a new method `fillOptions` in PropertyGroupOptionCollection to avoid the additional check, I did a little performance check and it reduces 20%-40% depends on the number of entities 

```php
    public function fillOptions(PropertyGroupOptionCollection $options): void
    {
        foreach ($options as $option) {
            $this->elements[$option->getUniqueIdentifier()] = $option;
        }
    }
```